pages/instructorCourseAdminAccess/instructorCourseAdminAccess.js

```diff
@@ -1,55 +1,292 @@
+                              + '<div class="container"><pre class="bg-dark text-white rounded p-2">'
+                              + result.not_given_cip.join(',\n')
+                              + '</pre></div>'
+                              + `<p>If you return to the <a href="${req.originalUrl}">access page</a>, you will find these `
+                              + `users in the list of course staff and can add student data access to each of them.</p>`;
+                }
+                if (result.not_given_cp.length > 0) {
+                    err.info += '<hr>'
+                              + '<p>The following users were <strong>not</strong> added to the course staff:</p>'
+                              + '<div class="container"><pre class="bg-dark text-white rounded p-2">'
+                              + result.not_given_cp.join(',\n')
+                              + '</pre></div>'
+                              + `<p>If you return to the <a href="${req.originalUrl}">access page</a>, you can try `
+                              + `to add them again. However, you should first check the reason for each failure to `
+                              + `grant access (see below). For example, it may be that a user you tried to add `
+                              + `was already a member of the course staff, in which case you will find them in the `
+                              + `list and can update their course content acccess as appropriate.</p>`;
+                }
+                err.info += '<hr>'
+                          + '<p>Here is the reason for each failure to grant access:</p>'
+                          + '<div class="container"><pre class="bg-dark text-white rounded p-2">'
+                          + result.errors.join('\n\n')
+                          + '</pre></div>';
+                return next(err);
+            }
+            res.redirect(req.originalUrl);
+        });
+    } else if (req.body.__action == 'course_permissions_insert_by_user_uid') {
+        let uid = req.body.uid.trim();
+        if (!uid) return next(error.make(400, `Empty UID`));
+
+        const params = [
             res.locals.course.id,
-            req.body.uid,
-            req.body.course_role,
+            uid,
+            'None',
             res.locals.authz_data.authn_user.user_id,
         ];
-        sqldb.call('course_permissions_insert_by_user_uid', params, function(err, _result) {
+        sqldb.call('course_permissions_insert_by_user_uid', params, (err, _result) => {
             if (ERR(err, next)) return;
             res.redirect(req.originalUrl);
         });
     } else if (req.body.__action == 'course_permissions_update_role') {
-        let params = [
+        if ((req.body.user_id == res.locals.user.user_id) && (!res.locals.authz_data.is_administrator)) {
+            return next(error.make(403, 'Owners cannot change their own course content access'));
```

Comment:
Yes, that's what it means.

The principle here is that no course should be without an Owner, because the course would then _require_ intervention by an administrator. The easiest way to achieve this is to not let an Owner remove themself as a course user or change their course role.

I believe the access page will encourage the current Owner to keep the list of users up to date. If I am wrong, this would be an easy change in a future PR.

(You could do something more complicated and allow an Owner to remove themself if there is at least one other Owner. I think that's unwise, because the other Owners may be long gone / inactive, but again - this sort of change would be easy to make in a future PR if it became clear that doing so was a good idea.)

Reply:
I'm happy to defer this for now. I do think we need an option to remove oneself at some point.